src/lgdo/types/histogram.py

```diff
@@ -0,0 +1,348 @@
+            super().__init__({"binedges": edges, "closedleft": Scalar(closedleft)})
+
+        @classmethod
+        def from_edges(
+            cls, edges: np.ndarray, binedge_attrs: dict[str, Any] | None = None
+        ) -> Histogram.Axis:
+            edge_diff = np.diff(edges)
+            if np.any(~np.isclose(edge_diff, edge_diff[0])):
+                return cls(edges, None, None, None, True, binedge_attrs)
+            return cls(None, edges[0], edges[-1], edge_diff[0], True, binedge_attrs)
+
+        @property
+        def is_range(self) -> bool:
+            return isinstance(self["binedges"], Struct)
+
+        @property
+        def first(self) -> float:
+            if not self.is_range:
+                msg = "Axis is not a range"
+                raise TypeError(msg)
+            return self["binedges"]["first"].value
+
+        @property
+        def last(self) -> float:
+            if not self.is_range:
+                msg = "Axis is not a range"
+                raise TypeError(msg)
+            return self["binedges"]["last"].value
+
+        @property
+        def step(self) -> float:
+            if not self.is_range:
+                msg = "Axis is not a range"
+                raise TypeError(msg)
+            return self["binedges"]["step"].value
+
+        @property
+        def closedleft(self) -> bool:
+            return self["closedleft"].value
+
+        @property
+        def nbins(self) -> int:
+            if self.is_range:
+                bins = (self.last - self.first) / self.step
+                bins_int = int(np.rint(bins))
+                assert np.isclose(bins, bins_int)
+                return bins_int
+            return len(self["binedges"].nda) - 1
+
+        @property
+        def edges(self) -> np.ndarray:
+            if self.is_range:
+                return np.linspace(self.first, self.last, self.nbins + 1)
+            return self["binedges"].nda
+
+        def __str__(self) -> str:
+            thr_orig = np.get_printoptions()["threshold"]
+            np.set_printoptions(threshold=8)
+
+            if self.is_range:
+                string = f"first={self.first}, last={self.last}, step={self.step}"
+            else:
+                string = f"edges={self.edges}"
+            string += f", closedleft={self.closedleft}"
+
+            attrs = self.get_binedgeattrs()
+            if attrs:
+                string += f" with attrs={attrs}"
+
+            np.set_printoptions(threshold=thr_orig)
+            return string
+
+        def get_binedgeattrs(self, datatype: bool = False) -> dict:
+            """Return a copy of the LGDO attributes dictionary of the binedges
+
+            Parameters
+            ----------
+            datatype
+                if ``False``, remove ``datatype`` attribute from the output
+                dictionary.
+            """
+            return self["binedges"].getattrs(datatype)
+
+    def __init__(
+        self,
+        weights: hist.Hist | np.ndarray,
```

Comment:
```suggestion
        weights: hist.Hist | NDArray,
```

we should use more generic types in function signatures (see [remarks here](https://learn.scientific-python.org/development/guides/mypy/#tips-for-good-types)). Could you quickly go through all your function signatures and fix this?